### ISSUE TITLE
Consensus handler should deliver messages to queues

### ISSUE THREAD

**Issue by jyellick** [hyperledger-archives/fabric#919]
Today, the consensus handler receives a message on a go-routine spawned by gRPC for a given connection.  The handler then uses this go routine to invoke the relevant piece of code, be it a Query, Transaction, Consensus, or Peer related function.

This pattern violates the Actor Model which OBC attempts to implement.  By consuming this thread for potentially unbounded amounts of time, it is possible to create deadlocks.  In particular, the interaction between state transfer (a peer message related function) and Sieve (a consensus message related function) has produced some rather inelegant workaround code (see issue #739 and PR #833), as this code attempts to release this message thread when state transfer might be needed.

After discussions with @jeffgarratt, the proper way to address this problem is to add a buffered channel for each message type.  In this way, each sub-component (Consensus, Peer, Query, and Exec) will not be stalled by the other.  This also allows prioritization; for instance queries are the least likely to be sensitive to failure, so the query buffer could be set to a much lower value than the the execution buffer.

The buffered channels should be of a configurable size, defined in the openchain.yaml peer section.  If a channel is full, the message should be discarded, after logging a warning and replying to the calling goroutine with an error.  Each channel should have an associated delivery side go routine which simply reads from the channel, and invokes the existing methods.  This should be a relatively small changeset, and can hopefully be confined strictly to `consensus/helper/handler.go`.

I'd like to invite comments especially from @corecode and @jeffgarratt, as well as @kchristidis, @tuand27613, and anyone else who is interested.

**Comment by kchristidis** [comment]
Following up on our discussion from a couple of days ago, I'm in favor of this change.

**Comment by tuand27613** [comment]
+1

Do we want to update the current code or integrate with what Marko/Simon/Chet are doing ?

**Comment by jyellick** [comment]
@tuand27613 It is always easier in my head than reality, but, I think that this is a sufficiently simple task that it is not a huge amount of time 'wasted', if it is not ported over.

**Comment by binhn** [comment]
@jyellick similarly, noops has a tx Q
BTW, one might be interested in the number of messages pending on the queue as a way to throttle message arrival

**Comment by jyellick** [comment]
@binhn Yes, I see that noops is doing this to some extent.  There is still the rebroadcast of a request which occurs on the message thread, and the channel write will block if the reading thread cannot keep up.

I have discussed a little with @corecode and @jeffgarratt and ultimately, the right way for this to work is to have the actors themselves hold the reading thread, with the channel in between the message thread and the actor thread.  However, in the interest of keeping this changeset relatively small, I'd like to do the intermediate step of simply delivering the messages in a non-blocking way, and having a handler specific go routine for each of the four noted message types.

I've just opened up a new issue ( #973 ) to track the more invasive change to an actor model, but will plan on implementing this simpler variation first, to more quickly solve outstanding bugs and to verify that there is currently no implicit reliance on the types of messages per peer not being processed concurrently.

**Comment by jyellick** [comment]
Per discussion with @jeffgarratt, @binhn, and @corecode  on Thursday, this will not make it into Sprint 3.

**Comment by corecode** [comment]
related to #1171

**Comment by jyellick** [comment]
This was implemented in #1325
